src/pages/Environments.js

```diff
@@ -116,8 +120,176 @@ const DeleteAppModal = ({ app: { googleProject, appName, diskName, appType }, on
   ])
 }
 
+const MigratePersistentDisksBanner = ({ count }) => {
+  const deadline = new Date('2023-01-01T00:00:00.000Z')
+  return div({
+    style: {
+      position: 'absolute', top: topBarHeight, left: '50%', transform: 'translate(-50%, -50%)',
+      backgroundColor: colors.warning(0.15),
+      border: '2px solid', borderRadius: '12px', borderColor: colors.warning(),
+      zIndex: 2 // Draw over top bar but behind contact support dialog
+    }
+  }, [
+    div({ style: { display: 'flex', alignItems: 'center', margin: '0.75rem 1.5rem 0.75rem 1.5rem' } }, [
+      icon('warning-standard', { size: 32, style: { color: colors.warning(), marginRight: '0.25rem' } }),
+      div([
+        strong([`You have ${Math.floor((deadline - Date.now()) / 86400000)} days to migrate ${count} shared persistent ${count > 1 ?
+          'disks' :
+          'disk'}. `]),
+        `Un-migrated disks will be DELETED after ${Utils.makeCompleteDate(deadline)}.`
+      ])
+    ])
+  ])
+}
+
+const MigratePersistentDiskCell = ({ onClick }) => div({
+  style: {
+    display: 'flex', flex: 1, flexDirection: 'column', textAlign: 'center',
+    height: '100%', margin: '-1rem', padding: '0.5rem 0 0 0',
+    backgroundColor: colors.danger(0.15), color: colors.danger()
+  }
+}, [
+  h(TooltipTrigger, { content: `This disk is shared between workspaces, which is no longer supported. Click "Migrate" to make copies for relevant workspaces.` }, [
+    div(['Offline', icon('warning-standard', { style: { marginLeft: '0.25rem', color: colors.danger() } })])
+  ]),
+  h(Link, { onClick, style: { wordBreak: 'break-word' } },
+    ['Migrate']
+  )
+])
+
+const MigratePersistentDiskModal = ({ disk, workspaces, onSuccess, onDismiss, onContactSupport, onDeleteDisk }) => {
+  // show a spinner when we're copying the disk to another workspace
+  const [migrating, setMigrating] = useState()
+  // workspaceId -> boolean indicating if we should copy the disk to the workspace
+  const [isWorkspaceSelected, setIsWorkspaceSelected] = useState({})
+  // users can choose to delete their disk instead of coping via a checkbox. Mutually exclusive with `isWorkspaceSelected`
+  const [deleteDisk, setDeleteDisk] = useState(false)
+
+  const copyDiskToWorkspace =
+    ({ googleProject, namespace: saturnWorkspaceNamespace, name: saturnWorkspaceName }) => {
+      // show an error for each failed copy operation
+      return reportErrorAndRethrow(`Error copying persistent disk to workspace '${saturnWorkspaceName}'`,
```

Comment:
Can we use `withErrorReportingInModal` here? Not too sure what the difference between that and this function is, but the one I mention handles some annoying things about error handling in modals, such as a modal obscuring the error based on positioning, and onDismiss/cleanup not properly being done after an error

Reply:
See https://github.com/DataBiosphere/terra-ui/pull/3301#discussion_r966031338